src/common/stack.rs

```diff
@@ -1,88 +1,419 @@
-		write!(f, "{}", self.0)
+	/// Create an new StackWithLimit with `limit` max size and `limit` elements pre-allocated
+	///
+	/// ```
+	/// # extern crate wasmi;
+	/// # use wasmi::{StackWithLimit, StackSize};
+	/// let bstack = StackWithLimit::<i32>::with_size(StackSize::from_element_count(2));
+	/// ```
+	pub fn with_size(size: StackSize<T>) -> StackWithLimit<T> {
+		StackWithLimit::new(StackSizeInitial(size), StackSizeLimit(size))
+	}
+
+	/// Attempt to push value onto stack.
+	///
+	/// # Errors
+	///
+	/// Returns Err(StackOverflow) if stack is already full.
+	#[inline]
+	pub(crate) fn push(&mut self, value: T) -> Result<(), StackOverflow> {
+		if self.stack.len() == self.stack.capacity() {
+			if self.stack.len() == self.limit {
+				return Err(StackOverflow);
+			}
+			// grows exponentially, just like Vec normally does
+			// https://doc.rust-lang.org/1.26.0/src/alloc/raw_vec.rs.html#462
+			let desired_len = self
+				.stack
+				.len()
+				.checked_mul(2)
+				.unwrap_or(usize::MAX)
+				.min(self.limit)
+				.max(1);
+			let additional_len = desired_len - self.stack.len();
+			self.stack.reserve_exact(additional_len);
+		}
+		debug_assert!(self.stack.len() < self.limit);
+		debug_assert!(self.stack.len() < self.stack.capacity());
+		let len = self.stack.len();
+		unsafe {
+			ptr::write(self.stack.get_unchecked_mut(len), value);
+			self.stack.set_len(len + 1);
+		}
+		Ok(())
+	}
+
+	pub(crate) fn pop(&mut self) -> Option<T> {
+		self.stack.pop()
+	}
+
+	/// Remove and Return top element. Does not check for emptyness.
+	/// If this is called on a zero length stack, bad things will happen.
+	/// Do not call this method unless you can prove the stack has length.
+	#[inline]
+	pub(crate) unsafe fn pop_unchecked(&mut self) -> T {
+		debug_assert!(self.stack.len() > 0);
+		let len = self.stack.len();
+		self.stack.set_len(len - 1);
+		ptr::read(self.stack.get_unchecked(self.stack.len()))
+	}
+
+	/// Return optional reference to item `depth` distance away from top
+	///
+	/// `bstack.get_relative_to_top(0)` gets the top of the stack
+	///
+	/// `bstack.get_relative_to_top(1)` gets the item just below the stack
+	pub(crate) fn get_relative_to_top(&self, depth: usize) -> Option<&T> {
+		// Be cognizant of integer underflow and overflow here. Both are possible in this situation.
+		// len() is unsigned, so if len() == 0, subtraction is a problem
+		// depth can legally be 2^32. On a 32 bit system, adding may overflow
+		// overflow isn't an issue unless T is zero size
+		// In debug builds, underflow panics, but in release mode, underflow is not checked.
+
+		let index = self.stack.len().checked_sub(1)?.checked_sub(depth)?;
+		debug_assert!(self.stack.len() > index, "guaranteed by previous line");
+		Some(&self.stack[index])
+	}
+
+	/// Return mutable reference to item `depth` distance away from top
+	///
+	/// Does not check whether depth is in range.
+	pub(crate) fn get_relative_to_top_mut_unchecked(&mut self, depth: usize) -> &mut T {
```

Comment:
See above, I'd suggest `nth_from_top_mut_unchecked`. Also, the name is misleading because it _is_ checked, it's just that it panicks if the depth is wrong. `unchecked` is almost exclusively reserved for `unsafe` functions.

Reply:
Changed it to `nth_from_top_mut_unchecked`.

Do you have a less misleading name in mind?

Reply:
Changed it to `nth_from_top_mut` and documented panic condition.